### PR TITLE
Lower composite tenstorrent.uniform to ttir.rand

### PR DESCRIPTION
**Pull request by umalesTT**
### Ticket
References https://github.com/tenstorrent/tt-mlir/issues/4854
Closes https://github.com/tenstorrent/tt-mlir/issues/5066

### Problem description
Details are written out in https://github.com/tenstorrent/tt-mlir/issues/4854

### What's changed
Added conversion for lowering composite tenstorrent.uniform op to ttir.rand.
Refactored a bit the tests for StableHLO composite to TTIR pass.

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp [hunk 1]
```diff
@@ -73,6 +73,65 @@ class StableHLOToTTIRCompositeOpConversionPattern
...
+    // Extract shape and convert to I32ArrayAttr.
+    auto arrayShapeAttr =
+        mlir::dyn_cast_or_null<ArrayAttr>(compositeAttrs.get("shape"));
+    SmallVector<int32_t> shapeI32;
+    for (auto attr : arrayShapeAttr) {
+      if (auto intAttr = mlir::dyn_cast<IntegerAttr>(attr)) {
+        shapeI32.push_back(static_cast<int32_t>(intAttr.getInt()));
+      }
+    }
+    auto sizeAttr = rewriter.getI32ArrayAttr(shapeI32);
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp]: I don't get this part, if `arrayShapeAttr` is not null, `sizeAttr` will end up being the same as `arrayShapeAttr`?

File: lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp [hunk 2]
```diff
@@ -73,6 +73,65 @@ class StableHLOToTTIRCompositeOpConversionPattern
...
+
+    // Extract low and high from constant operands.
+    auto lowOp =
+        adaptor.getOperands()[1].getDefiningOp<mlir::stablehlo::ConstantOp>();
+    auto highOp =
+        adaptor.getOperands()[2].getDefiningOp<mlir::stablehlo::ConstantOp>();
+
+    auto lowValue = mlir::dyn_cast<DenseFPElementsAttr>(lowOp.getValue());
+    auto highValue = mlir::dyn_cast<DenseFPElementsAttr>(highOp.getValue());
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp]: Doing `dyn_cast` (`getDefiningOp` also has `dyn_cast` under the hood) without subseqent checking doesn't make sense. Either do `cast`, or at least assert after each `dyn_cast`, `assert` error is usually much more readable than nullptr dereference.

File: lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp [hunk 3]
```diff
@@ -73,6 +73,65 @@ class StableHLOToTTIRCompositeOpConversionPattern
...
+
+    auto lowAttr = rewriter.getF32FloatAttr(lowValue.getValues<float>()[0]);
+    auto highAttr = rewriter.getF32FloatAttr(highValue.getValues<float>()[0]);
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp:121]: Assert that size is one here.

File: test/ttmlir/Conversion/StableHLOToTTIR/composite/test_uniform.mlir [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    %cst_17 = stablehlo.constant dense<-0.00417768164> : tensor<50xf32>
+    %cst_18 = stablehlo.constant dense<0.00943887047> : tensor<50xf32>
+    %cst_19 = stablehlo.constant dense<0.246640727> : tensor<50xf32>
+    %cst_20 = stablehlo.constant dense<1.00167406> : tensor<50xf32>
+    %cst_21 = stablehlo.constant dense<1.50140941> : tensor<50xf32>
+    %cst_22 = stablehlo.constant dense<2.83297682> : tensor<50xf32>
+    %cst_23 = stablehlo.constant dense<1.000000e+00> : tensor<50xf32>
+    %cst_24 = stablehlo.constant dense<0x7F800000> : tensor<50xf32>
+    // CHECK: ttir.rand
+    // CHECK-NOT: stablehlo.bitcast_convert
```
Comment (azecevic000) [inline test/ttmlir/Conversion/StableHLOToTTIR/composite/test_uniform.mlir]: If you want to check that `bitcast_convert` isn't in the graph, the best way might be
```
// CHECK-NOT: stablehlo.bitcast_convert
// CHECK: ttir.rand
// CHECK-SAME: dtype = f32, high = 1.000000e+00 : f32, low = -0.99999994 : f32, seed = 0 : ui32, size = [50 : i32]
// CHECK-NOT: stablehlo.bitcast_convert
```

File: lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp [hunk 5]
```diff
@@ -73,6 +73,65 @@ class StableHLOToTTIRCompositeOpConversionPattern
...
+  matchAndRewrite(mlir::stablehlo::CompositeOp srcOp,
+                  mlir::stablehlo::CompositeOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (srcOp.getName() != "tenstorrent.uniform") {
+      return failure();
+    }
+
+    auto outputType =
+        mlir::cast<RankedTensorType>(srcOp.getResult(0).getType());
+    auto compositeAttrs = srcOp.getCompositeAttributes();
```
Comment (anuragsingh-tt) [inline lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp]: Let's not use `auto` here as its not clear what the data type is.

File: lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp (hunk 2 above)
Comment (anuragsingh-tt) [reply lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp]: Yes, let's use the `rewriter.notifyMatchFailure` here.

**Comment by anuragsingh-tt** [comment]
It's still not clear to me reading the linked PR what the `tenstorrent.uniform` op is. Can you provide some more details?

File: lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp (hunk 1 above)
Comment (umalesTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp]: This was for purpose of ArrayAttr->I32ArrayAttr conversion, as ttir.rand requires I32 for shape attribute.
Instead of this conversion in MLIR, I made sure on frontend level that emitted shape is of type I32 so no need for this anymore.

File: lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp (hunk 2 above)
Comment (umalesTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp]: Doing ```cast``` now with assert that ```DefiningOp``` is indeed ```ConstantOp```.

File: test/ttmlir/Conversion/StableHLOToTTIR/composite/test_uniform.mlir (hunk 4 above)
Comment (umalesTT) [reply test/ttmlir/Conversion/StableHLOToTTIR/composite/test_uniform.mlir]: Check for ```stablehlo.bitcast_convert``` was not needed, so deleted it

File: lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp (hunk 5 above)
Comment (umalesTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp]: Using ```DictionaryAttr``` now

**Comment by umalesTT** [comment]
> It's still not clear to me reading the linked PR what the `tenstorrent.uniform` op is. Can you provide some more details?

Yes, sure. So inbuilt dropout function in Jax emits 300+ StableHLO ops, where vast majority of them come from generating uniform distribution via ```jax.random.uniform```. In the body of this uniform function in StableHLO ```stablehlo.bitcast_convert``` is emitted which we don't support yet. 
So if we could manage to lower uniform distribution generation part of StableHLO right away to ```ttir.rand``` we would circumvent a) yet unsupported ops b) managing 200+ ```ttir``` ops with just one. 
The way to do this is to monkeypatch function ```jax.random.uniform``` in Jax and enforce it to emit composite op ```tenstorrent.uniform``` which then, by this PR, we just translate to ```ttir.rand```.

The monkeypatch of ```tenstorrent.uniform``` is well-tested locally, however it is still not present on tt-xla repo, because without this MLIR support the conversion would fail right away. Once this PR is merged, ```tenstorrent.uniform``` will be on PR and merged in tt-xla.

File: lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp [hunk 6]
```diff
@@ -73,6 +74,67 @@ class StableHLOToTTIRCompositeOpConversionPattern
...
+    }
+
+    auto outputType =
+        mlir::cast<RankedTensorType>(srcOp.getResult(0).getType());
+
+    DictionaryAttr compositeAttrs = srcOp.getCompositeAttributes();
+
+    // Extract shape attribute.
+    auto sizeAttr =
+        mlir::dyn_cast_or_null<ArrayAttr>(compositeAttrs.get("shape"));
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp]: You can do just `mlir::cast` here, if there isn't `shape` attr or it's not `ArrayAttr` we want this to fail immediately instead of failing during `ttir.rand` op generation.

Comment (umalesTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp]: Yeah sure, done.

**Comment by jojo-771771** [comment]
I can work on this. Please assign!

**Comment by jojo-771771** [comment]
I can work on this. Please assign!

**Comment by jojo-771771** [comment]
I can work on this. Please assign!
